pkg/backup/actions/csi/pvc_action.go

```diff
@@ -570,3 +581,485 @@ func NewPvcBackupItemAction(f client.Factory) plugincommon.HandlerInitializer {
+		}
+
+		if existingVS != nil {
+			if existingVS.Status != nil && existingVS.Status.VolumeGroupSnapshotName != nil {
+				p.log.Infof("Reusing existing VolumeSnapshot %s for PVC %s", existingVS.Name, pvc.Name)
+				return existingVS, nil
+			} else {
+				return nil, errors.Errorf("found VolumeSnapshot %s for PVC %s, but it was not created via VGS (missing volumeGroupSnapshotName)", existingVS.Name, pvc.Name)
+			}
+		}
+
+		p.log.Infof("No existing VS found for PVC %s; creating new VGS", pvc.Name)
+
+		// List all PVCs in the VGS group
+		groupedPVCs, err := p.listGroupedPVCs(pvc.Namespace, vgsLabelKey, group)
+		if err != nil {
+			return nil, err
+		}
+
+		// Determine the CSI driver for the grouped PVCs
+		driver, err := p.determineCSIDriver(groupedPVCs)
+		if err != nil {
+			return nil, err
+		}
+		if driver == "" {
+			return nil, errors.New("multiple CSI drivers found for grouped PVCs; failing backup")
+		}
+
+		// Determine the VGSClass to be used for the VGS object to be created
+		vgsClass, err := p.determineVGSClass(driver, backup, &pvc)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create the VGS object
+		newVGS, err := p.createVolumeGroupSnapshot(backup, pvc, vgsLabelKey, group, vgsClass)
+		if err != nil {
+			return nil, err
+		}
+
+		// Wait for all the VS objects associated with the VGS to have status and VGS Name (VS readiness is checked in legacy flow)
+		err = p.waitForVGSAssociatedVS(ctx, newVGS, backup.Spec.CSISnapshotTimeout.Duration)
+		if err != nil {
+			return nil, err
+		}
+
+		// Update the VS objects: remove VGS owner references and finalizers; add backup metadata labels.
+		err = p.updateVGSCreatedVS(newVGS, backup)
+		if err != nil {
+			return nil, err
+		}
+
+		// Wait for VGSC binding in the VGS status
+		err = p.waitForVGSCBinding(ctx, newVGS, backup.Spec.CSISnapshotTimeout.Duration)
+		if err != nil {
+			return nil, err
+		}
+
+		// Patch the VGSC deletionPolicy to Retain.
+		err = p.patchVGSCDeletionPolicy(newVGS)
+		if err != nil {
+			return nil, err
+		}
+
+		// Delete the VGS and VGSC
+		err = p.deleteVGSAndVGSC(newVGS)
+		if err != nil {
+			return nil, err
+		}
+
+		// Fetch the VS that was created for this PVC via VGS.
+		vs, err := p.getVSForPVC(ctx, pvc.Name, pvc.Namespace, newVGS.Name, backup)
+		if err != nil {
+			return nil, err
+		}
+
+		return vs, nil
+	}
+
+	// Legacy fallback: create individual VS
+	return p.createVolumeSnapshot(pvc, backup)
+}
+
+func (p *pvcBackupItemAction) findExistingVSForBackup(
+	backupUID types.UID,
+	backupName, pvcName, namespace string,
+) (*snapshotv1api.VolumeSnapshot, error) {
+	vsList := &snapshotv1api.VolumeSnapshotList{}
+	if err := p.crClient.List(context.TODO(), vsList, crclient.InNamespace(namespace)); err != nil {
+		return nil, errors.Wrap(err, "failed to list VolumeSnapshots")
+	}
+
+	for _, vs := range vsList.Items {
+		if vs.Labels == nil {
+			continue
+		}
+
+		if vs.Labels[velerov1api.BackupNameLabel] == label.GetValidName(backupName) &&
```

Comment:
Let's put the label selection as a selector in `crClient.List`, so as to get better performance in scaled case